lib/command.js

```diff
@@ -412,6 +412,27 @@ Expecting one of '${allowedValues.join("', '")}'`);
     return this;
   }
 
+  /**
+   * Add hook to await argument and option values before calling action handlers for this command and its nested subcommands.
+   * Useful for asynchronous custom processing (parseArg) of arguments and option-arguments.
+   */
+
+  awaitHook() {
+    this.hook('preAction', async function awaitHook(thisCommand, actionCommand) {
+      actionCommand.processedArgs = await Promise.all(
+        actionCommand.processedArgs
+      );
+
+      for (const [key, value] of Object.entries(actionCommand.opts())) {
+        actionCommand.setOptionValueWithSource(
+          key, await value, actionCommand._optionValueSources[key]
```

Comment:
1) The option value may be an array of values for a variadic option.

2) I think each call to await will delay execution until the next tick. I suggest only call for `thenable` values.

3) I would slightly prefer not to call `setOptionValueWithSource` on every value, which also fits in with testing for `thenable` in above item.

Reply:
There is an existing test for `thenable` here: https://github.com/tj/commander.js/blob/4ef19faac1564743d8c7e3ce89ef8d190e1551b4/lib/command.js#L1192

Reply:
1. Due to the "previous value" paradigm of custom variadic option processing, the only possible sources for such an array are default and implied values which are assumed to be preprocessed and should ideally be left untouched. If the array really needs to be awaited elementwise, one could simply use `Promise.all(arr)` as the default / implied value.
2. Solved by [4738e3c](https://github.com/tj/commander.js/pull/1902/commits/4738e3c3ea63ce8c02f85915d13b5de6d31c1df7).
3. Solved by [4738e3c](https://github.com/tj/commander.js/pull/1902/commits/4738e3c3ea63ce8c02f85915d13b5de6d31c1df7).